modules/vectorfieldvisualizationgl/glsl/lic2d.frag

```diff
@@ -24,88 +24,128 @@
+#  define KERNEL_NORMALIZATION
+#  endif
+#endif
+#if !defined(INTEGRATION_DIRECTION)
+#  define INTEGRATION_DIRECTION 0
+#endif
+
 uniform sampler2D inport;
+uniform ImageParameters inportParameters;
+
 uniform sampler2D noiseTexture;
 
 uniform int samples;
 uniform float stepLength;
-uniform bool normalizeVectors;
+uniform mat2 invBasis = mat2(1);
+
 uniform bool useRK4;
 
 uniform bool postProcessing = false;
 uniform bool intensityMapping = false;
-uniform float brightness = 0.0; // in [-1, 1]
-uniform float contrast = 0.0;   // in [-1, 1]
+uniform float brightness = 0.0;  // in [-1, 1]
+uniform float contrast = 0.0;    // in [-1, 1]
 uniform float gamma = 1.0;
 
 in vec3 texCoord_;
 
+vec2 euler(in vec2 pos) {
+    vec2 v = getValueTexel(inport, inportParameters, pos).xy;
+#if defined(NORMALIZATION)
+    v = normalize(v);
+#endif
+    return v;
+}
 
-vec2 euler(vec2 posF ){
-    vec2 V0 = texture(inport, posF).rg;
-    if(normalizeVectors){
-        V0 = normalize(V0);
-    }
-    return V0;
+vec2 rungekutta4(in vec2 p0, in vec2 v0, float stepsize) {
+    vec2 p1 = p0 + v0 * stepsize * 0.5;
+    vec2 v1 = euler(p1);
 
-}
- 
-vec2 rk4(vec2 p0 , float stepsize) {
-    vec2 V0 = euler(p0);
-    
-    vec2 p1 = p0 + V0 * stepsize/2;
-    vec2 V1 = euler(p1);
-    
-    vec2 p2 = p0 + V1 * stepsize/2;
-    vec2 V2 = euler(p2);
+    vec2 p2 = p0 + v1 * stepsize * 0.5;
+    vec2 v2 = euler(p2);
 
-    vec2 p3 = p0 + V2 * stepsize;
-    vec2 V3 = euler(p3);
+    vec2 p3 = p0 + v2 * stepsize;
+    vec2 v3 = euler(p3);
 
+    return (v0 + 2.0 * (v1 + v2) + v3) / 6.0;
+}
 
-    return (V0 + 2*(V1+V2) + V3) / 6.0;
+bool insideUnitSquare(in vec2 pos) {
+    return pos == clamp(pos, vec2(0), vec2(1));
 }
 
-void traverse(inout  float v , inout int c,vec2 posF , float stepSize,int steps){
-    for(int i = 0;i<steps;i++){
-        vec2 V0;
-        if(useRK4){
-            V0 = rk4(posF,stepSize);
-        }else{
-            V0 = euler(posF);
-        }
-        posF += V0 * stepSize;
+float box(in int step) {
+    return 1.0;
+}
 
-        if(posF.x < 0 ) break;
-        if(posF.y < 0 ) break;
+float gaussian(in int step) {
+    float weight = 2.0 / float(samples);
+#if INTEGRATION_DIRECTION == 0
+    weight *= 0.5;
+#endif
 
-        if(posF.x > 1 ) break;
-        if(posF.y > 1 ) break;
+    float dist = step / float(samples);
+    float distSq = dot(dist, dist);
+    float kernelSizeSq = 1.0 / 9.0;
+    return clamp(exp(-0.693147180559945 * distSq / kernelSizeSq) * weight, 0.0, 1.0);
+}
 
-        v += texture(noiseTexture, posF.xy).r;
-        c += 1;
+float integration(in vec2 pos, in int steps, in float stepSize, inout int stepCount) {
+    float density = 0.0;
+    int integrationSteps = 1;
+    for (int i = 0; i < steps; ++i) {
+        vec2 v = euler(pos);
+#if defined(USE_RUNGEKUTTA)
+        v = rungekutta4(pos, v, stepSize);
+#endif
+        v *= stepSize;
+        pos += invBasis * v;
+        // TODO: consider texture repeat?
```

Comment:
could just have a uniform for the clamp ranges (0, and 1) and pass in min/max for the directions that does not have clamping...